src/types/CodepointWidthDetector_gen.go

```diff
@@ -0,0 +1,566 @@
+	if len(os.Args) <= 1 {
+		fmt.Println(`Usage:
+    go run CodepointWidthDetector_gen.go <path to ucd.nounihan.grouped.xml>
+
+You can download the latest ucd.nounihan.grouped.xml from:
+    https://www.unicode.org/Public/UCD/latest/ucdxml/ucd.nounihan.grouped.zip`)
+		os.Exit(1)
+	}
+
+	data, err := os.ReadFile(os.Args[1])
+	if err != nil {
+		return fmt.Errorf("failed to read XML: %w", err)
+	}
+
+	ucd := &UCD{}
+	err = xml.Unmarshal(data, ucd)
+	if err != nil {
+		return fmt.Errorf("failed to parse XML: %w", err)
+	}
+
+	values, err := extractValuesFromUCD(ucd)
+	if err != nil {
+		return err
+	}
+
+	// More stages = Less size. The trajectory roughly follows a+b*c^stages, where c < 1.
+	// 4 still gives ~30% savings over 3 stages and going beyond 5 gives diminishing returns (<10%).
+	trie := buildBestTrie(values, 2, 8, 4)
+	rules := buildJoinRules()
+
+	for cp, expected := range values {
+		var v TrieType
+		for _, s := range trie.Stages {
+			v = s.Values[int(v)+((cp>>s.Shift)&s.Mask)]
+		}
+		if v != expected {
+			return fmt.Errorf("trie sanity check failed for %U", cp)
+		}
+	}
+
+	buf := &strings.Builder{}
+
+	_, _ = fmt.Fprintf(buf, "// Generated by CodepointWidthDetector_gen.go\n")
+	_, _ = fmt.Fprintf(buf, "// on %v, from %s, %d bytes\n", time.Now().UTC().Format(time.RFC3339), ucd.Description, trie.TotalSize+len(rules)*len(rules))
+	_, _ = fmt.Fprintf(buf, "// clang-format off\n")
+
+	for i, s := range trie.Stages {
+		width := 16
+		if i != 0 {
+			width = s.Mask + 1
+		}
+		_, _ = fmt.Fprintf(buf, "static constexpr uint%d_t s_stage%d[] = {", s.Bits, i+1)
+		for j, value := range s.Values {
+			if j%width == 0 {
+				buf.WriteString("\n   ")
+			}
+			_, _ = fmt.Fprintf(buf, " 0x%0*x,", s.Bits/4, value)
+		}
+		buf.WriteString("\n};\n")
+	}
+
+	_, _ = fmt.Fprintf(buf, "static constexpr uint8_t s_joinRules[%d][%d] = {", len(rules), len(rules))
+	for _, row := range rules {
+		buf.WriteString("\n   ")
+		for _, val := range row {
+			var i int
+			if val {
+				i = 1
+			}
+			_, _ = fmt.Fprintf(buf, " %d,", i)
+		}
+	}
+	buf.WriteString("\n};\n")
+
+	_, _ = fmt.Fprintf(buf, "[[msvc::forceinline]] constexpr uint%d_t ucdLookup(const char32_t cp) noexcept\n", trie.Stages[len(trie.Stages)-1].Bits)
+	buf.WriteString("{\n")
+	for i, s := range trie.Stages {
+		_, _ = fmt.Fprintf(buf, "    const auto s%d = s_stage%d[", i+1, i+1)
+		if i == 0 {
+			_, _ = fmt.Fprintf(buf, "cp >> %d", s.Shift)
+		} else {
+			_, _ = fmt.Fprintf(buf, "s%d + ((cp >> %d) & %d)", i, s.Shift, s.Mask)
+		}
+		buf.WriteString("];\n")
+	}
+	_, _ = fmt.Fprintf(buf, "    return s%d;\n", len(trie.Stages))
+	buf.WriteString("}\n")
+
+	buf.WriteString("[[msvc::forceinline]] constexpr uint8_t ucdGraphemeJoins(const uint8_t lead, const uint8_t trail) noexcept\n")
+	buf.WriteString("{\n")
+	buf.WriteString("    const auto l = lead & 15;\n")
+	buf.WriteString("    const auto t = trail & 15;\n")
+	buf.WriteString("    return s_joinRules[l][t];\n")
+	buf.WriteString("}\n")

```

Comment:
Why did we generate this as opposed to make it static code in the cpp file? Is it so that the Generator can change the lookup code when the table layout changes?

Reply:
If so: it might be easier to read/write if given as a multiline string...

```go
buf.WriteString(`
[[msvc::forceinline]] ...
`)
```